File: mx_driving/common/ops/csrc/Hypot.cpp
// Copyright (c) 2024 Huawei Technologies Co., Ltd
// All rights reserved.

#include "csrc/OpApiCommon.h"
#include "functions.h"

at::Tensor npu_hypot(const at::Tensor& x, const at::Tensor& y)
{
    auto out = at::empty_like(x, x.options());
    EXEC_NPU_CMD(aclnnHypot, x, y, out);
    return out;
}

std::tuple<at::Tensor, at::Tensor> npu_hypot_grad(const at::Tensor& x, const at::Tensor& y, const at::Tensor& out, const at::Tensor& out_grad)
{
    auto x_grad = at::empty_like(x, x.options());
    auto y_grad = at::empty_like(y, y.options());
    EXEC_NPU_CMD(aclnnHypotGrad, x, y, out, out_grad, x_grad, y_grad);
    return std::make_tuple(x_grad, y_grad);
}

File: mx_driving/common/ops/hypot.py
"""
Copyright (c) Huawei Technologies Co., Ltd. 2024. All rights reserved.
"""
import torch
from torch.autograd import Function

import torch_npu
import ads_c


class Hypot(Function):
    @staticmethod
    def forward(ctx, x, y):
        x_broadcasted, y_broadcasted = torch.broadcast_tensors(x, y)
        out = ads_c.npu_hypot(x_broadcasted.contiguous(), y_broadcasted.contiguous())
        ctx.save_for_backward(x, y, out);
        return out

    @staticmethod
    def backward(ctx, out_grad):
        x, y, out = ctx.saved_tensors
        x_broadcasted, y_broadcasted = torch.broadcast_tensors(x, y)
        x_grad, y_grad = ads_c.npu_hypot_grad(x_broadcasted.contiguous(), y_broadcasted.contiguous(), out, out_grad)

        # reshape the broadcasted tensors to origin tensors and sum the grad
        for dim, size in enumerate(x.shape):
            if size == 1:
                x_grad = x_grad.sum(dim, keepdim=True)
        for dim, size in enumerate(y.shape):
            if size == 1:
                y_grad = y_grad.sum(dim, keepdim=True)

        return x_grad, y_grad

hypot = Hypot.apply
